py3status/modules/speedtest.py

```diff
@@ -0,0 +1,313 @@
+        self.thresholds_init = self.py3.get_color_names_list(self.format)
+        self.command = "speedtest-cli --json --secure"
+        self.url = None
+
+        self.quality = {
+            -1: "failed",
+            0: "unknow",
+            1: "bad",
+            2: "good",
+            3: "slower",
+            4: "faster",
+        }
+
+        # if download* or upload* missing, run complete test
+        if any(
+            key.startswith(x)
+            for x in ["download", "upload"]
+            for key in self.placeholders
+        ):
+            if not any(key.startswith("upload") for key in self.placeholders):
+                self.command += " --no-upload"
+
+            if not any(key.startswith("download") for key in self.placeholders):
+                self.command += " --no-download"
+
+        if self.server_id and int(self.server_id):
+            self.command += " --server " + str(self.server_id)
+
+        if self.button_share and all(
+            x in self.placeholders for x in ["download", "upload"]
+        ):
+            self.command += " --share"
+
+        self.command += " --timeout " + str(self.timeout)
+
+    def _is_running(self):
+        try:
+            self.py3.command_output(["pgrep", "speedtest-cli"])
+            return True
+        except self.py3.CommandError:
+            return False
+
+    def _get_speedtest_data(self):
+        return loads(self.py3.command_output(self.command)) or None
+
+    def _start(self):
+        self.cached_until = self.py3.time_in(self.sleep_timeout)
+
+        if not self._is_running():
+            self.cached_until = self.py3.CACHE_FOREVER
+            previous_data = self.py3.storage_get("speedtest_data") or {}
+            current_data = self._get_speedtest_data() or {}
+
+            if current_data and len(current_data) > 1:
+                # create a "total" for know if cnx is better or not
+                # between two run
+                current_data["total"] = int(current_data.get("download", 0)) + int(
+                    current_data.get("upload", 0)
+                )
+
+                # create "quality" #maybe bad name
+                if previous_data and "total" in previous_data:
+                    if current_data["total"] >= previous_data["total"]:
+                        quality_key = 4
+                    else:
+                        quality_key = 3
+                else:
+                    if current_data["total"] <= 0:
+                        quality_key = 1
+                    else:
+                        quality_key = 2
+                current_data["quality"] = quality_key
+                current_data["quality_name"] = self.quality[quality_key]
+
+        # zero-ing if not fetched, raw version and units convertion
+        for x in ["download", "upload", "bytes_received", "bytes_sent"]:
+            unit = self.unit_size if "bytes" in x else self.unit_bitrate
+            current_data[x] = current_data.get(x, 0)
```

Comment:
We don't have to run `self.py3.format_units()` and other things all times on all four placeholders... Only what we need from `format`. It might be good to skip the rest of work if we get `None`.